ext/drbbdup/drbbdup.c

```diff
@@ -708,29 +697,51 @@ drbbdup_encode_runtime_case(void *drcontext, drbbdup_per_thread *pt, void *tag,
     }
 
     /* Encoding is application-specific and therefore we need to user to define the
-     * encoding of the runtime case. We invoke a user-defined call-back.
-     */
-    ASSERT(opts.insert_encode, "The encode call-back cannot be NULL");
-    /* Note, we could tell the user not to reserve flags and scratch register since
-     * drbbdup is doing that already. However, for flexibility/backwards compatibility
-     * ease, this might not be the best approach.
+     * encoding of the runtime case. Therefore, we invoke a user-defined call-back.
+     *
+     * It could also be the encoding is done directly and changed on demand. Therefore,
+     * the call-back may be NULL.
      */
-    opts.insert_encode(drcontext, tag, bb, where, opts.user_data, pt->orig_analysis_data);
+    if (opts.insert_encode != NULL) {
 
-    /* Restore all unreserved registers used by the call-back. */
-    drreg_restore_all(drcontext, bb, where);
+        /* Note, we could tell the user not to reserve flags and scratch register since
+         * drbbdup is doing that already. However, for flexibility/backwards compatibility
+         * ease, this might not be the best approach.
+         */
+        opts.insert_encode(drcontext, tag, bb, where, opts.user_data,
+                           pt->orig_analysis_data);
 
-#ifdef X86_32
-    /* Load the encoding to the scratch register.
-     * The dispatcher could compare directly via mem, but this will
-     * destroy micro-fusing (mem and immed).
-     */
+        /* Restore all unreserved registers used by the call-back. */
+        drreg_restore_all(drcontext, bb, where);
+    }
+
+    /* Load the encoding to the scratch register. */
     opnd_t scratch_reg_opnd = opnd_create_reg(DRBBDUP_SCRATCH_REG);
-    opnd_t opnd = drbbdup_get_encoding_opnd();
-    instr_t *instr = INSTR_CREATE_mov_ld(drcontext, scratch_reg_opnd, opnd);
+    instr_t *instr =
+        INSTR_CREATE_mov_ld(drcontext, scratch_reg_opnd, opts.runtime_case_opnd);
+    instrlist_meta_preinsert(bb, where, instr);
+}
+
+#ifdef X86_64
+static void
+drbbdup_insert_compare_encoding(void *drcontext, instrlist_t *bb, instr_t *where,
+                                drbbdup_case_t *current_case, reg_id_t reg_encoding)
+{
+    opnd_t opnd = opnd_create_abs_addr(&current_case->encoding, OPSZ_PTR);
```

Comment:
This seems like a bug: `drbbdup_manager_t.cases` is *not* allocated as reachable heap, so it may not be reachable (i.e., more than 2GB away from the code cache).

Reply:
Thanks for the catch. I'm afraid I am going to have to make it reachable then because otherwise the dispatcher will need to spill an additional register (which I want to avoid).

Reply:
Is dr_global_alloc() non reachable heap memory please?

Docs seem this is the case: 

> Memory allocated through dr_thread_alloc(), dr_global_alloc(), and dr_nonheap_alloc() is also reachable

https://dynamorio.org/dynamorio_docs/dr__tools_8h.html#abb7e20e83f7b8e1b65428e45c8ab4211

Reply:
Hmm, I thought it was all un-reachable by default, which is definitely the case for heap allocated inside DR, where you have to call a special routine to get reachable heap.  Looks like I made all the client interface stuff reachable by default, though.  Sorry, there's no problem here then.